rust/src/extensions.rs

```diff
@@ -0,0 +1,191 @@
+//! Core logic for extensions.yaml file.
+
+/*
+ * Copyright (C) 2020 Red Hat, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ */
+
+use anyhow::{bail, Context, Result};
+use openat_ext::OpenatDirExt;
+use serde_derive::Deserialize;
+use std::collections::HashMap;
+use std::io::Read;
+
+use crate::ffi::StringMapping;
+use crate::utils;
+
+const RPMOSTREE_EXTENSIONS_STATE_FILE: &str = ".rpm-ostree-state-chksum";
+
+#[derive(Deserialize, Debug)]
+pub struct Extensions {
+    extensions: HashMap<String, Extension>,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Extension {
+    packages: Vec<String>,
+    architectures: Option<Vec<String>>,
+    #[serde(rename = "match-base-evr")]
+    match_base_evr: Option<String>,
+}
+
+fn extensions_load_stream(
+    stream: &mut impl std::io::Read,
+    basearch: &str,
+    base_pkgs: &Vec<StringMapping>,
+) -> Result<Box<Extensions>> {
+    let mut parsed: Extensions = serde_yaml::from_reader(stream)?;
+
+    parsed.extensions.retain(|_, ext| {
+        ext.architectures
+            .as_ref()
+            .map(|v| v.iter().any(|a| a == basearch))
+            .unwrap_or(true)
+    });
+
+    let base_pkgs: HashMap<String, String> =
+        base_pkgs.iter().cloned().map(|i| (i.k, i.v)).collect();
+
+    for (_, ext) in parsed.extensions.iter_mut() {
+        for pkg in &ext.packages {
+            if base_pkgs.contains_key(pkg) {
+                bail!("package {} already present in base", pkg);
+            }
+        }
+        if let Some(ref matched_base_pkg) = ext.match_base_evr {
+            let evr = base_pkgs
+                .get(matched_base_pkg)
+                .with_context(|| format!("couldn't find base package {}", matched_base_pkg))?;
+            let pkgs = ext
+                .packages
+                .iter()
+                .map(|pkg| format!("{}-{}", pkg, evr))
+                .collect();
+            ext.packages = pkgs;
+        }
+    }
+
+    Ok(Box::new(parsed))
+}
+
+pub(crate) fn extensions_load(
+    path: &str,
+    basearch: &str,
+    base_pkgs: &Vec<StringMapping>,
+) -> Result<Box<Extensions>> {
+    let f = utils::open_file(path)?;
+    let mut f = std::io::BufReader::new(f);
+    extensions_load_stream(&mut f, basearch, base_pkgs).with_context(|| format!("parsing {}", path))
+}
+
+impl Extensions {
+    pub(crate) fn get_packages(&self) -> Vec<String> {
+        self.extensions
+            .iter()
+            .flat_map(|(_, ext)| ext.packages.iter().cloned())
+            .collect()
+    }
+
+    pub(crate) fn state_checksum_changed(&self, chksum: &str, output_dir: &str) -> Result<bool> {
```

Comment:
Anything exposed via cxx currently needs to go via `CxxResult`, see https://github.com/coreos/rpm-ostree/pull/2451

Hmm maybe this would be less of a trap if we specified `CxxResult` in `lib.rs` too...

Reply:
(I tried that, doesn't work)